lib/core/src/Cardano/Wallet/Primitive/Slotting.hs

```diff
@@ -66,31 +78,100 @@ import Numeric.Natural
     ( Summary (..), neverForksSummary )
 
 import qualified Cardano.Slotting.Slot as Cardano
 import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
+import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
+import qualified Ouroboros.Consensus.HardFork.History.Summary as HF
 
 -- -----------------------------------------------------------------------------
 -- New Api using ouroboros-consensus. With the right interpreter, the
 -- calculations don't break on hard-forks.
 
+
+--
+-- Queries
+--
+
 epochOf :: Cardano.SlotNo -> Qry EpochNo
-epochOf slot = do
-    (e, _, _) <- slotToEpoch slot
+epochOf slot = HardForkQry $ do
+    (e, _, _) <- HF.slotToEpoch slot
     return $ EpochNo $ fromIntegral $ Cardano.unEpochNo e
 
+startTime :: Cardano.SlotNo -> Qry UTCTime
+startTime s = do
+    rel <- HardForkQry (fst <$> HF.slotToWallclock s)
+    RelToUTCTime rel
+
+slotId :: EpochNo -> SlotInEpoch -> Qry Cardano.SlotNo
+slotId = error "TODO#1901: implement slotId :: EpochNo -> SlotInEpoch -> SlotNo"
+
+slotRangeFromTimeRange
+    :: Range UTCTime
+    -> Qry (Maybe (Range Cardano.SlotNo))
+slotRangeFromTimeRange (Range Nothing Nothing) = pure $ Just wholeRange
+slotRangeFromTimeRange (Range low hi) = do
+    low' <- liftMay slotNoCeiling low
+    hi' <- liftMay slotAtTime' hi
+    pure $ Range low' <$> (maybe (Just Nothing) (fmap Just) hi')
+  where
+    slotAtTime' :: UTCTime -> Qry (Maybe Cardano.SlotNo)
+    slotAtTime' x = do
+         slotAtTime x >>= \case
+            Just (slot, _timeInSlot, _timeRemainingInSlot) -> pure $ Just slot
+            Nothing -> pure Nothing
+
+    slotAtTime t = do
+        UTCTimeToRel t >>= \case
+            Just relTime -> fmap Just $ HardForkQry $ HF.wallclockToSlot relTime
+            Nothing -> return Nothing
+
+    liftMay :: (a -> Qry b) -> Maybe a -> Qry (Maybe b)
+    liftMay f (Just x) = Just <$> f x
+    liftMay _ Nothing = return Nothing
+
+    -- NOTE: /Perhaps/ it would be nicer to treat Qry and Maybe in the same
+    -- monad. I didn't get around to do it though since I don't know if it
+    -- actually will work out.
+    slotNoCeiling :: UTCTime -> Qry Cardano.SlotNo
+    slotNoCeiling t = do
+         slotAtTime t >>= \case
+            Just (_, _, remaining) -> do
+                slotAtTime (remaining `addUTCTime` t) >>= \case
+                    Just (s, _, _) ->
+                        return s
+                    Nothing ->
+                        return $ Cardano.SlotNo 0
+            Nothing -> do
+                return $ Cardano.SlotNo 0
+
+-- A @TimeInterpreter@ is a way for the wallet to run things of type @Qry a@.
+--
+-- NOTE:
+-- A @TimeInterpreter@ could in theory decide to update the era summary from the
+-- node when running a query.
+--
+-- We cannot manually specify when the fetching happens.
+--
+-- This may or may not be what we actually want.
+type TimeInterpreter m = forall a. Qry a -> m a
```

Comment:
☝️ We may now have a `TimeInterpreter IO` that _sometimes_ takes a long time to run a conversion query (when it needs to refetch the summary over LSQ).

The re-fetching should happen on very rare occasions though…

Reply:
I think it would be better to have the network layer query the Interpreter on a schedule, and stash it in a mutable variable. So that all queries can be pure.

Reply:
Let's see. I was thinking that the initial Interpreter will be one constructed from genesis parameters.
It will get updated once the node has synced far enough. (Question - how far is enough?)
This would prevent the wallet from being unresponsive while the node is syncing.